File: REFERENCE_SNAPSHOT_v32.72/github_source/deploy/hf/drive_doc_identity.py
"""One Drive slot per document kind in a patient folder.

PDF, baked validation video, original analysis clip, overlay JSON, and
kinematics JSON each keep a single canonical name. Re-export updates that
file in place; leftover dated / alias names of the same kind are trash.
"""
from __future__ import annotations

import re
from pathlib import Path
from typing import Optional, Tuple

_PHASES = ("pre", "post", "healthy", "baseline")


def _sanitize(name: str) -> str:
    return re.sub(r"[^\w.\-]", "_", (name or "").strip())[:180]


def _norm_phase(phase: str) -> str:
    p = (phase or "").strip().lower()
    if p in ("baseline", "healthy", "healthy_side"):
        return "healthy"
    if p in ("pre", "post"):
        return p
    return ""


def _norm_stem(value: str) -> str:
    return re.sub(r"[^a-z0-9]+", "", (value or "").lower())


def _file_person_stem(filename: str) -> str:
    """Name portion of ``{digits}_{Name}.pdf``; empty when the file is untagged."""
    stem = Path(filename or "").stem
    head, _, rest = stem.partition("_")
    if head.isdigit() and rest:
        return rest
    return ""


def _folder_person_stem(folder_key: str) -> str:
    stem = Path(folder_key or "").stem
    head, _, rest = stem.partition("_")
    if head.isdigit() and rest:
        return rest
    return stem


def _basename(name: str) -> str:
    return Path((name or "").replace("\\", "/").split("/")[-1]).name


def document_kind(name: str) -> Optional[str]:
    """Stable slot id: clinic_report, pre_validation, pre_validation_original, …"""
    raw = _basename(name)
    if not raw:
        return None
    lower = raw.lower()
    stem = Path(lower).stem

    if lower.endswith(".pdf"):
        return "clinic_report"

    if lower.endswith(".json"):
        for phase in _PHASES:
            np = _norm_phase(phase)
            if stem in (f"{phase}_validation_overlay", f"{np}_validation_overlay"):
                return f"{np}_validation_overlay"
            if stem in (f"{phase}_kinematics", f"{np}_kinematics"):
                return f"{np}_kinematics"
        if stem.endswith("_validation_overlay"):
            np = _norm_phase(stem[: -len("_validation_overlay")])
            return f"{np}_validation_overlay" if np else None
        if stem.endswith("_kinematics"):
            np = _norm_phase(stem[: -len("_kinematics")])
            return f"{np}_kinematics" if np else None
        return None

    if not lower.endswith((".mp4", ".mov", ".m4v", ".webm")):
        return None

    if stem.endswith("_validation_original") or (
        stem.endswith("_original") and not stem.endswith("_validation_original")
    ):
        base = stem[: -len("_validation_original")] if stem.endswith("_validation_original") else stem[: -len("_original")]
        np = _norm_phase(base)
        return f"{np}_validation_original" if np else None

    for phase in _PHASES:
        np = _norm_phase(phase)
        aliases = (
            f"{phase}_validation",
            f"{phase}_validation_unified",
            f"{phase}_unified_validation",
            f"{np}_validation",
            f"{np}_validation_unified",
            f"{np}_unified_validation",
        )
        if stem in aliases:
            return f"{np}_validation"
    return None


def canonical_name(
    kind: str,
    patient_key: str = "",
    uploaded_name: str = "",
) -> Optional[str]:
    kind = (kind or "").strip().lower()
    if not kind:
        return None
    if kind == "clinic_report":
        key = _sanitize(patient_key)[:120]
        if key:
            return f"{key}.pdf"
        stem = Path(_basename(uploaded_name)).stem
        low = stem.lower()
        if low.startswith("report_") or low in ("clinic_report", "report"):
            return "clinic_report.pdf"
        return f"{_sanitize(stem) or 'report'}.pdf"
    if kind.endswith("_validation_original"):
        phase = kind[: -len("_validation_original")]
        return f"{phase}_validation_original.mp4"
    if kind.endswith("_validation_overlay"):
        phase = kind[: -len("_validation_overlay")]
        return f"{phase}_validation_overlay.json"
    if kind.endswith("_kinematics"):
        phase = kind[: -len("_kinematics")]
        return f"{phase}_kinematics.json"
    if kind.endswith("_validation"):
        phase = kind[: -len("_validation")]
        return f"{phase}_validation.mp4"
    return None


def canonicalize_upload_name(name: str, patient_key: str = "") -> Optional[str]:
    """Map any alias / dated export name onto the single Drive slot for that kind."""
    kind = document_kind(name)
    if not kind:
        return None
    return canonical_name(kind, patient_key=patient_key, uploaded_name=name)


def is_foreign_person_pdf(filename: str, folder_key: str) -> bool:
    """True when a PDF is tagged as a different person than this patient folder."""
    if not str(filename or "").lower().endswith(".pdf"):
        return False
    file_stem = _file_person_stem(filename)
    if not file_stem:
        return False
    want = _norm_stem(_folder_person_stem(folder_key))
    if not want or len(want) < 3:
        return False
    return _norm_stem(file_stem) != want


def should_trash_as_alias(
    child_name: str,
    keep_name: str,
    *,
    folder_key: str = "",
    keep_id: str = "",
    child_id: str = "",
) -> bool:
    if keep_id and child_id and keep_id == child_id:
        return False
    child = _basename(child_name)
    keep = _basename(keep_name)
    if not child or not keep:
        return False
    kind = document_kind(keep)
    if not kind or document_kind(child) != kind:
        return False
    if kind == "clinic_report" and is_foreign_person_pdf(child, folder_key):
        return False
    if child == keep:
        return bool(keep_id and child_id and child_id != keep_id)
    return True


def same_kind_alias_examples(kind: str, patient_key: str = "") -> Tuple[str, ...]:
    """Human-facing leftover names that collapse into ``kind`` (tests / restore)."""
    kind = (kind or "").strip().lower()
    if kind == "clinic_report":
        key = _sanitize(patient_key)[:120] or "clinic_report"
        return (
            f"{key}.pdf",
            "clinic_report.pdf",
            "report.pdf",
            "report_115_patient_2026-09-14.pdf",
        )
    if kind.endswith("_validation_original"):
        phase = kind[: -len("_validation_original")]
        return (
            f"{phase}_validation_original.mp4",
            f"{phase}_original.mp4",
        )
    if kind.endswith("_validation_overlay"):
        phase = kind[: -len("_validation_overlay")]
        return (f"{phase}_validation_overlay.json",)
    if kind.endswith("_kinematics"):
        phase = kind[: -len("_kinematics")]
        return (f"{phase}_kinematics.json",)
    if kind.endswith("_validation"):
        phase = kind[: -len("_validation")]
        return (
            f"{phase}_validation.mp4",
            f"{phase}_validation.webm",
            f"{phase}_validation_unified.mp4",
            f"{phase}_validation_unified.webm",
        )
    return ()

File: REFERENCE_SNAPSHOT_v32.97/github_source/frontend/src/driveDocIdentity.js
/**
 * One Google Drive slot per document kind in a patient folder.
 * Re-export updates that file; dated / alias names of the same kind are leftovers.
 */

const PHASES = ["pre", "post", "healthy", "baseline"];

function sanitize(name) {
  return String(name || "").trim().replace(/[^\w.\-]/g, "_").slice(0, 180);
}

function basename(name) {
  return String(name || "").replace(/\\/g, "/").split("/").pop() || "";
}

function normPhase(phase) {
  const p = String(phase || "").trim().toLowerCase();
  if (p === "baseline" || p === "healthy" || p === "healthy_side") return "healthy";
  if (p === "pre" || p === "post") return p;
  return "";
}

function stemOf(filename) {
  const base = basename(filename);
  const i = base.lastIndexOf(".");
  return (i > 0 ? base.slice(0, i) : base).toLowerCase();
}

export function documentKind(name) {
  const raw = basename(name);
  if (!raw) return null;
  const lower = raw.toLowerCase();
  const stem = stemOf(lower);

  if (lower.endsWith(".pdf")) return "clinic_report";

  if (lower.endsWith(".json")) {
    for (const phase of PHASES) {
      const np = normPhase(phase);
      if (stem === `${phase}_validation_overlay` || stem === `${np}_validation_overlay`) {
        return `${np}_validation_overlay`;
      }
      if (stem === `${phase}_kinematics` || stem === `${np}_kinematics`) {
        return `${np}_kinematics`;
      }
    }
    return null;
  }

  if (!/\.(mp4|mov|m4v|webm)$/i.test(lower)) return null;

  if (stem.endsWith("_validation_original") || (stem.endsWith("_original") && !stem.endsWith("_validation_original"))) {
    const base = stem.endsWith("_validation_original")
      ? stem.slice(0, -"_validation_original".length)
      : stem.slice(0, -"_original".length);
    const np = normPhase(base);
    return np ? `${np}_validation_original` : null;
  }

  for (const phase of PHASES) {
    const np = normPhase(phase);
    const aliases = [
      `${phase}_validation`,
      `${phase}_validation_unified`,
      `${phase}_unified_validation`,
      `${np}_validation`,
      `${np}_validation_unified`,
      `${np}_unified_validation`,
    ];
    if (aliases.includes(stem)) return `${np}_validation`;
  }
  return null;
}

export function canonicalDriveName(name, patientKey = "") {
  const kind = documentKind(name);
  if (!kind) return null;
  if (kind === "clinic_report") {
    const key = sanitize(patientKey).slice(0, 120);
    if (key) return `${key}.pdf`;
    const stem = stemOf(name);
    if (stem.startsWith("report_") || stem === "clinic_report" || stem === "report") {
      return "clinic_report.pdf";
    }
    return `${sanitize(stem) || "report"}.pdf`;
  }
  if (kind.endsWith("_validation_original")) {
    return `${kind.slice(0, -"_validation_original".length)}_validation_original.mp4`;
  }
  if (kind.endsWith("_validation_overlay")) {
    return `${kind.slice(0, -"_validation_overlay".length)}_validation_overlay.json`;
  }
  if (kind.endsWith("_kinematics")) {
    return `${kind.slice(0, -"_kinematics".length)}_kinematics.json`;
  }
  if (kind.endsWith("_validation")) {
    return `${kind.slice(0, -"_validation".length)}_validation.mp4`;
  }
  return null;
}

/** Stable Drive PDF name for this patient. Local download can still use a readable label. */
export function clinicReportDriveName(patientKey) {
  const key = sanitize(patientKey).slice(0, 120);
  return key ? `${key}.pdf` : "clinic_report.pdf";
}

/** Same folder key the clinic app uses when backing up a patient. */
export function patientDriveKeyFromDemographics(demographics, fallbackId) {
  const d = demographics || {};
  const id = String(d.participantId || fallbackId || "").trim();
  const name = String(d.name || d.fullName || "").trim();
  if (!id && !name) return "";
  const slug = (s) => String(s).replace(/[^\w.\-]/g, "_").slice(0, 100);
  if (id && name) return slug(`${id}_${name}`);
  return slug(id || name);
}

export function driveNameCandidates(primaryName) {
  const names = [primaryName];
  const mapped = canonicalDriveName(primaryName);
  if (mapped) names.push(mapped);
  const kind = documentKind(primaryName);
  if (kind === "clinic_report") {
    names.push("clinic_report.pdf", "report.pdf");
  } else if (kind && kind.endsWith("_validation_original")) {
    const phase = kind.slice(0, -"_validation_original".length);
    names.push(`${phase}_validation_original.mp4`, `${phase}_original.mp4`);
  } else if (kind && kind.endsWith("_validation") && !kind.includes("original") && !kind.includes("overlay")) {
    const phase = kind.slice(0, -"_validation".length);
    names.push(
      `${phase}_validation.mp4`,
      `${phase}_validation.webm`,
      `${phase}_validation_unified.mp4`,
      `${phase}_validation_unified.webm`,
    );
  }
  return [...new Set(names.filter(Boolean))];
}
